lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,341 @@
+      selectCaseIntRangeVals.push_back(intRangeValue);
+    } else {
+      ReportOverlapCaseValError(
+          std::get<2>(*selectCaseIntRangeValsIter), source);
+    }
+  }
+  void Insert(std::optional<std::string> lowerVal,
+      std::optional<std::string> upperVal, parser::CharBlock source) {
+    const auto &stringRangeValue{std::make_tuple(lowerVal, upperVal, source)};
+    selectCaseValsIter = std::find_if(selectCaseVals.begin(),
+        selectCaseVals.end(), [&stringRangeValue](const auto &mem) {
+          return (
+              (!std::get<0>(stringRangeValue) ||
+                  ((std::get<0>(stringRangeValue).value())
+                          .compare(std::get<std::string>(mem.first)) >= 0)) &&
+              (!std::get<1>(stringRangeValue) ||
+                  ((std::get<1>(stringRangeValue).value())
+                          .compare(std::get<std::string>(mem.first)) <= 0)));
+        });
+    if (selectCaseValsIter != selectCaseVals.end()) {
+      ReportOverlapCaseValError(selectCaseValsIter->second, source);
+    }
+
+    selectCaseStringRangeValsIter = std::find_if(
+        selectCaseStringRangeVals.begin(), selectCaseStringRangeVals.end(),
+        [&stringRangeValue](const auto &mem) {
+          return ((!std::get<0>(mem) && !std::get<0>(stringRangeValue)) ||
+              (!std::get<1>(mem) && !std::get<1>(stringRangeValue)) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<0>(stringRangeValue) &&
+                       ((std::get<0>(stringRangeValue).value())
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<0>(stringRangeValue) &&
+                          ((std::get<0>(stringRangeValue).value())
+                                  .compare(std::get<1>(mem).value()) <= 0)))) ||
+              ((!std::get<0>(mem) ||
+                   (std::get<1>(stringRangeValue) &&
+                       (std::get<1>(stringRangeValue)
+                               .value()
+                               .compare(std::get<0>(mem).value()) >= 0))) &&
+                  (!std::get<1>(mem) ||
+                      (std::get<1>(stringRangeValue) &&
+                          (std::get<1>(stringRangeValue)
+                                  .value()
+                                  .compare(std::get<1>(mem).value()) <= 0)))));
+        });
+
+    if (selectCaseStringRangeValsIter == selectCaseStringRangeVals.end()) {
+      selectCaseStringRangeVals.push_back(stringRangeValue);
+    } else {
+      ReportOverlapCaseValError(
+          std::get<2>(*selectCaseStringRangeValsIter), source);
+    }
+  }
+
+private:
+  using selectCaseStmtTypes =
+      std::vector<std::pair<std::variant<std::int64_t, bool, std::string>,
+          parser::CharBlock>>;
+  selectCaseStmtTypes selectCaseVals;
+  selectCaseStmtTypes::iterator selectCaseValsIter;
```

Comment:
What is this for?

Reply:
Since case expression value can be either a case value  or case value range.
To store case value of one of the three allowed types (along with the location info), Either we can have separate a vector of pairs (value, location) for each type, or vector of pairs of variant of 3 types and the source location. So I choose the later option.
Any better approach to save these?